### ISSUE TITLE
How to do ensemble prediction?

### ISSUE THREAD

**Issue by ghost** [titu1994/Snapshot-Ensembles#5]
I have a model like this

```
#create model
model = Sequential()
model.add(Dense(5, input_dim=5, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(3, activation='sigmoid'))
model.compile(loss='mse', optimizer='adam', metrics=['accuracy'])
```

I have used this pseudo code to train and  create 5 ensemble models.

```
from snapshot import SnapshotCallbackBuilder

M = 5 # number of snapshots
nb_epoch = T = 200 # number of epochs
alpha_zero = 0.1 # initial learning rate
model_prefix = 'Model_'

snapshot = SnapshotCallbackBuilder(T, M, alpha_zero) 
model.fit(trainX, trainY, batch_size=50,callbacks=snapshot.get_callbacks(model_prefix=model_prefix))
```

it created 5 external *hdf5* files now how should I load and use this for prediction? Can you give a simple example?

**Comment by titu1994** [comment]
See the predict cifar 10 or predict cifar 100 scripts. They show how to do ensemble prediction.

Edit : Removed the above comment. Seems you are doing a regression task with 3 possible outputs and mean squared error as loss.

In that case the predict scripts wont help much..

**Comment by ghost** [comment]
I already seen it but the code is too complicated I only understand this part
```
train_preds = []
for fn in models_filenames:
    model.load_weights(fn)
    print("Predicting train set values on model %s" % (fn))
    yPreds = model.predict(trainX, batch_size=128, verbose=2)
    train_preds.append(yPreds)
```
we are loading the 5 models and do prediction then append those predictions. What we should do after that? I am looking for a simplified code, a pseudo code like in your readme do the job.

> Seems you are doing a regression task with 3 possible outputs and mean squared error as loss.
> In that case the predict scripts wont help much..

Please forgive me but can I ask why regression task wont work? and also I am squasing my predictions between 0 and 1 so my predicted value will not go to infinity.

**Comment by titu1994** [comment]
I meant the script as it is won't work. Of course your model will work, provided it trains properly. I am assuming you are doing multi task learning if you have multiply outputs of sigmoid, but the same concept applies. 

Once you have a list of 5 predictions like above, create a numpy array using `preds = np.array(train_preds)`.

After that do `final_predictions = np.mean(preds, axis=0)`

That is your final prediction vector of shape (nb of samples, 3)

**Comment by ghost** [comment]
Thanks for the quick reply. Can I ask a last question What is calculate_weighted_accuracy? what is its purpose and use?

```
def calculate_weighted_accuracy():
    global weighted_predictions, weight, prediction, yPred, yTrue, accuracy, error
    weighted_predictions = np.zeros((testX.shape[0], nb_classes), dtype='float32')
    for weight, prediction in zip(prediction_weights, test_preds):
        weighted_predictions += weight * prediction
    yPred = np.argmax(weighted_predictions, axis=1)
    yTrue = testY
    accuracy = metrics.accuracy_score(yTrue, yPred) * 100
    error = 100 - accuracy
    print("Accuracy : ", accuracy)
    print("Error : ", error)
    exit()

```

**Comment by titu1994** [comment]
It was a mistake. The idea was to use cross validation metrics to create weights for each of the models, but I didn't implement it properly. 

I'll have to remove that sometime.
